src/lib/optimizer/strategy/dependent_group_by_reduction_rule.cpp

```diff
@@ -24,155 +24,103 @@ using namespace opossum;  // NOLINT
-      // unique or primary key) in hope that the shorter one will later form the group-by clause.
-      std::vector<std::pair<size_t, size_t>> constraints_position_and_size;
-      constraints_position_and_size.reserve(table_constraints.size());
-      auto constraint_id = size_t{0};
-      for (const auto& table_constraint : table_constraints) {
-        constraints_position_and_size.emplace_back(constraint_id, table_constraint.columns.size());
-        ++constraint_id;
-      }
-      std::sort(constraints_position_and_size.begin(), constraints_position_and_size.end(),
-                [](const auto& left, const auto& right) { return left.second < right.second; });
-
-      // Try to reduce the group-by list one constraint at a time, starting with the shortest constraint. Multiple
-      // constraints (e.g., a primary key and a unique column constraint) might result in different group-by lists
-      // (depending on the number of columns of the constraint), but we stop as soon as one constraint successfully
-      // reduced the group-by list. The reason is that there is no advantage in a second reduction if the first
-      // reduction was successful, because no further columns will be removed. Hence, as soon as one reduction took
-      // place, we can ignore the remaining constraints.
-      for (const auto& [position, size] : constraints_position_and_size) {
-        const auto& table_constraint = table_constraints[position];
-        group_by_list_changed |= reduce_group_by_columns_for_constraint(table_constraint, group_by_columns,
-                                                                        stored_table_node, aggregate_node);
-        if (group_by_list_changed) break;
-      }
+    // Gather group-by columns
+    ExpressionUnorderedSet group_by_columns(aggregate_node.aggregate_expressions_begin_idx + 1);
+    auto node_expressions_iter = aggregate_node.node_expressions.cbegin();
+    std::copy(node_expressions_iter, node_expressions_iter + aggregate_node.aggregate_expressions_begin_idx,
+              std::inserter(group_by_columns, group_by_columns.end()));
+
+    // Sort the FDs by their left set's column count in hope that the shortest will later form the group-by clause.
+    std::sort(fds.begin(), fds.end(),
+              [](const auto& fd_left, const auto& fd_right) { return fd_left.first.size() < fd_right.first.size(); });
+
+    // --- Main ---
+    // Try to reduce the group-by list one constraint at a time, starting with the shortest constraint. Multiple
+    // constraints might result in different group-by lists (depending on the number of columns of the constraint).
+    // We stop as soon as one constraint successfully reduced the group-by list. There is no advantage in a second
+    // reduction if the first reduction was successful, because no further columns will be removed. Hence, as soon as
+    // one reduction took place, we can ignore the remaining constraints.
```

Comment:
You didn't add this, but is it actually true? Thinking of a table with column a,b,c,d,e, FDs `a->b`, `c->d` (Mr. Codd, please forgive me), and `GROUP BY a,b,c,d`, the rule might actually be applied twice.

Reply:
Sounds right. That comment is only true for PKs, not for general FKs. At least your examples sounds rights to me. Can we distinguish between FKs from PKs and from other "sources"?
If not, it should rarely matter if we have multiple reductions runs even though the not-first runs might no longer reduce the list.